lgc/builder/ImageBuilder.cpp

```diff
@@ -1317,8 +1318,24 @@ Value *ImageBuilder::CreateImageQuerySamples(unsigned dim, unsigned flags, Value
   // Check if resource type is 2D MSAA or 2D MSAA array, 14 = SQ_RSRC_IMG_2D_MSAA, 15 = SQ_RSRC_IMG_2D_MSAA_ARRAY
   Value *isMsaa = CreateOr(CreateICmpEQ(imageType, getInt32(14)), CreateICmpEQ(imageType, getInt32(15)));
 
-  // Return sample number if resource type is 2D MSAA or 2D MSAA array. Otherwise, return 1.
-  return CreateSelect(isMsaa, sampleNumber, getInt32(1), instName);
+  // Sample number is 1 if resource type is neither 2D MSAA nor 2D MSAA array.
+  sampleNumber = CreateSelect(isMsaa, sampleNumber, getInt32(1), instName);
+
+  // The sampler number should be 0 if allowNullDescriptor is on and image descriptor is a null descriptor
+  if (m_pipelineState->getOptions().allowNullDescriptor) {
+    // Check if the image descriptor all bits 0
+    Value *const nullDescCmp = CreateICmpEQ(imageDesc, Constant::getNullValue(imageDesc->getType()));
+    Value *isNullDesc = CreateExtractElement(nullDescCmp, uint64_t(0));
+    const unsigned elemCount =
+        m_pipelineState->getLgcContext()->getTargetInfo().getGpuProperty().descriptorSizeResource / sizeof(unsigned);
+    for (unsigned elemIdx = 1; elemIdx < elemCount; ++elemIdx) {
+      Value *elem = CreateExtractElement(nullDescCmp, elemIdx);
+      isNullDesc = CreateAnd(isNullDesc, elem);
+    }
+    sampleNumber = CreateSelect(isNullDesc, getInt32(0), sampleNumber);
+  }
```

Comment:
We can do better here than looping over the entire descriptor.

On all supported hardware generations, the TYPE field (bits 127:124, or the top 4 bits of dword 3) are in the range 8 to 15 for valid image descriptors. So it's sufficient to check dword 3 against 0.

Reply:
I checked the internal implementation of get sampler info. So I changed the logic to return sample number or 0 by verifying DATA_FORMAT field of dword1 whether it is 0. Please review.